### DISCUSSION TITLE
Only want to transfer Y (from YUV420)

### DISCUSSION THREAD

**Discussion by Hornochse007** [raspberrypi/picamera2#438] [Q&A]
Hello everybody, 
 
  I think, this would be a very easy question, but I don't find the answer. Maybe anybody can help me? 

I  need only the Y-Plane from the buffer, but this really fast. So if I try to cut the np.array, the time will be to long. My idea was to use a new configuration.
Instead of the config {'format':'YUV420', 'size':(4056,3040), 'stride':4096, 'framesize':18677760} 
I want a new format: e.g. {'format':'Y', 'size':(4056,3040), 'stride':4064, 'framesize':12354560} or the yuv420 and there only 3040 rows. 

`
from picamera2 import Picamera2

pc2 = Picamera2()
cc = pc2.create_still_configuration(main={"size":(4056,3040), "format":"YUV420"}}
pc2.configure(cc)
pc2.start()
image = pc2.capture_array("main")
pc2.stop()
`

Maybe anybody has an idea for me? 

Thanks a lot and stay healthy, 
   Hornochse007

**Comment by davidplowman** [comment]
Hi, can you just do
```
Y = image[:3040, :4056]
```
I think that will just give you a "view" of the data without copying it, so it should be fast. (Some operations want proper contiguous data, in which case you would have to copy it, it will depend what you want to do.)

**Reply by Hornochse007** [reply]
Hi, 
  thank you for the quick answer. 
I tried this before, but I always have additional 200 ms for this process (only for the view). So it was my idea to read out only the Y from the buffer, so that i will have less data to read and transfer and so it will speed up. 

Sorry for my bad english.

**Reply by davidplowman** [reply]
Interesting - can you say a bit more how you have determined that it takes 200ms? When I tried this:
```
    start = time.monotonic()
    Y = image[:3040, :4056]
    diff = time.monotonic() - start
```
it recorded a time almost always between 2 and 3ms (this is on a Pi 4).

**Reply by Hornochse007** [reply]
Crazy,  i will have a look at the time, i am on a Pi4, 2GB.
Feedback will come soon.

**Reply by Hornochse007** [reply]
Ok, it's my fault. You are right. 
It's not complete my code, but i want to optimize the times. 
I found the additional time in another area, where the image will be transfered to PIL:
`
from PIL import Image
image = Image.fromarray(Y)
 `
This behaviour is strange: 
- if i put in the complete image, it will be 0.001s,
- if i put in the smaller view, i will have 300ms. 

But i think this is not the correct discussion for this problem. 


By the way, i am just on the way to grab only the Y on the buffer, to avoid data, Do you think there would be possible to set the stride and framesize? 

Thanks a lot

**Reply by davidplowman** [reply]
Yes, it's all to do with whether the pixel data is contiguous or not. Many algorithms/libraries work happily with non-contiguous data, but others do not. Unfortunately non-contiguous data can be difficult to avoid when dealing with certain hardware devices. The hardware in the Pi can only write out images with a line-to-line stride that is a multiple of 64 bytes, so there's no way you can have a stride of 4056. However, I have some other ideas for you.

Firstly, I found that this:
```
    Y = image[:3040, :4056]
    Y = np.ascontiguousarray(Y)
    im = Image.fromarray(Y)
```
takes only 45ms, not 300. So when PIL has to do the conversion to contiguous pixels, it seems to be doing it in a bad way.

But perhaps a better idea is to ask for a slightly differently sized image, so that the output can be contiguous. For example, use `"size": (4032, 3024)` in your config. This will give you a contiguous buffer, and
```
    Y = image[:3024, :4032]
    im = Image.fromarray(Y)
```
appears to take under 3ms again.

**Reply by Hornochse007** [reply]
>     Y = image[:3040, :4056]
>     Y = np.ascontiguousarray(Y)
>     im = Image.fromarray(Y)
This skips the nearly the time from the conversion to PIL (max. 60ms)! Fantastic!! Thanks a lot! 

Last Step: I would try to reduce the size of the readout of the buffer to only the Y-plane. Do you think i have a chance to change this parameters somehow? I tried to set the stride and framesize but couldn't do it. If I can reduce it the framesize to 2/3 of the YUV [so only Y - with the resolution, i think my time will shrink into 2/3 the time from buffer readout.

**Reply by davidplowman** [reply]
Yes, I'm afraid you can't choose your own stride, the system chooses the minimum valid one for you. Also there is no "Y only" format, and adding such a thing would be quite involved and require changing at least libcamera, but possibly also kernel drivers and maybe even firmware on the GPU (which is for most practical purposes impossible).

You could try using `capture_buffer` instead of `capture_array` and then using `reshape`. I don't know if that would be faster. I would certainly recommend using a size like (4032, 3024) if you can because that does make some things much faster.

**Reply by Hornochse007** [reply]
Ok, i will try and send a feedback.

**Reply by davidplowman** [reply]
Any feedback about a feature like that would come back to Raspberry Pi, so I think this is probably the right place to discuss it. However, given that it's a bit tricky and we have the YUV420 format, I don't think it would be regarded as a priority. Might there be other things we can do to help your code go faster? For example, you get can a low resolution image as well as the full resolution image and in some applications these can help.

**Reply by Hornochse007** [reply]
Sorry, i didn't get time to test, i will do it next week and give a feedback. 

And where should i place the request for the Y-readout ? Here or in libcam? 

I need only the high resolution to work with. 
Thank you for your help and have a nice weekend!
